### PR TITLE
storage: bugfixes

### PR DESCRIPTION
**Pull request by brauner**

### CODE REVIEW

File: lxd/storage.go [hunk 1]
```diff
@@ -714,7 +714,17 @@ func (ss *storageShared) setUnprivUserAcl(c container, destPath string) error {
 func (ss *storageShared) createImageDbPoolVolume(fingerprint string) error {
 	// Create a db entry for the storage volume of the image.
 	volumeConfig := map[string]string{}
-	_, err := dbStoragePoolVolumeCreate(ss.d.db, fingerprint, storagePoolVolumeTypeImage, ss.poolID, volumeConfig)
+	err := storageVolumeValidateConfig(ss.pool.Name, volumeConfig, ss.pool)
+	if err != nil {
+		return err
+	}
+
+	err = storageVolumeFillDefault(ss.pool.Name, volumeConfig, ss.pool)
```
Comment (brauner) [inline lxd/storage.go]:
```
chb@conventiont|~
> lxc storage volume show pool1 image/ac373f5bda6c7ad4d53267bba6644fa8bf2bdf1d0e6b7714549d668d597f88ba
name: ac373f5bda6c7ad4d53267bba6644fa8bf2bdf1d0e6b7714549d668d597f88ba
config:
  block.filesystem: ext4
  size: 25MB
type: image
used_by: []
```

File: lxc/storage.go [hunk 2]
```diff
@@ -646,6 +646,11 @@ func (c *storageCmd) doStoragePoolShow(client *lxd.Client, name string) error {
 		pool.Config["size"] = shared.GetByteSizeString(int64(sz), 0)
 	}
 
+	sz, err = strconv.ParseUint(pool.Config["volume.size"], 10, 64)
+	if err == nil {
+		pool.Config["volume.size"] = shared.GetByteSizeString(int64(sz), 0)
+	}
```
Comment (brauner) [inline lxc/storage.go]:
```
chb@conventiont|~
> lxc storage show pool1
config:
  lvm.vg_name: pool1
  source: pool1
  volume.size: 25MB
name: pool1
driver: lvm
used_by:
- /1.0/containers/alp1
- /1.0/images/ac373f5bda6c7ad4d53267bba6644fa8bf2bdf1d0e6b7714549d668d597f88ba
```

**Comment by brauner** [comment]
After an `LVM` upgrade now:

```
chb@conventiont|~
> lxc storage show blub
config:
  lvm.thinpool_name: LXDPool
  lvm.vg_name: blub
  source: blub
  volume.size: 15GB
name: blub
driver: lvm
used_by:
- /1.0/containers/ac373f5bda6c7ad4d53267bba6644fa8bf2bdf1d0e6b7714549d668d597f88ba
- /1.0/containers/alp1
- /1.0/containers/alp1/snapshots/snap0
- /1.0/containers/alp1/snapshots/snap1
- /1.0/containers/alp1/snapshots/snap2
- /1.0/containers/alp2
```

container:
```
chb@conventiont|~
> lxc storage volume show blub container/alp1
name: alp1
config:
  block.filesystem: ext4
  size: 10GB
type: container
used_by: []
```

image:

```
chb@conventiont|~
> lxc storage volume show blub image/ac373f5bda6c7ad4d53267bba6644fa8bf2bdf1d0e6b7714549d668d597f88ba
name: ac373f5bda6c7ad4d53267bba6644fa8bf2bdf1d0e6b7714549d668d597f88ba
config:
  block.filesystem: ext4
  size: 10GB
type: image
used_by: []
```

File: lxc/storage.go (hunk 2 above)
Comment (stgraber) [inline lxc/storage.go]: Hmm, so what we do for the container keys is that we store the value as it's provided by the user rather than convert it to bytes. We just attempt to convert it to bytes when received to validate it.

And we convert it to bytes whenever we actually need the value as bytes.

That way you don't get into the case where I set something as "10000MB" and have it returned to me as "10GB" by LXD.

File: lxd/patches.go [hunk 3]
```diff
@@ -1465,6 +1713,13 @@ func patchStorageApiKeys(name string, d *Daemon) error {
 		if pool.Driver == "zfs" {
 			pool.Config["zfs.pool_name"] = pool.Config["source"]
 		} else if pool.Driver == "lvm" {
+			// On previous upgrade versions, "size" was set instead
+			// of "volume.size", so transfer the value and then
+			// unset it.
+			if pool.Config["size"] != "" {
+				pool.Config["volume.size"] = pool.Config["size"]
+				pool.Config["size"] = ""
+			}
```
Comment (stgraber) [inline lxd/patches.go]: Shouldn't that be another patch so that anyone who's running git master will get the fix too?

Comment (brauner) [inline lxd/patches.go]: There will be another patch but I think it's also worth fixing this for users that run the upgrade for the first time.

Comment (stgraber) [inline lxd/patches.go]: If they run the upgrade for the first time, they won't have the "size" property set since they'll be running a fixed LXD that won't do that.

**Comment by stgraber** [comment]
So you really need to be careful with patches and DB schema updates. The only time where it's fine to change an old patch or schema update function is if you know that the bug you're fixing would have prevented the user from having that patch/schema update be marked as done in the database.

Otherwise, you're adding code that won't fix people who are running the latest release or running LXD from git (such as from the edge channel for snaps users).

File: lxd/patches.go (hunk 3 above)
Comment (brauner) [inline lxd/patches.go]: Well, it doesn't hurt also they might have performed a partial upgrade.

**Comment by brauner** [comment]
So people who have gone through that patch won't run that patch again. People who have performed a partial upgrade will benefit from this and as I said: there needs to be an additional patch anyway that deals with falsely set config key. This one will fix git master. Does that sound reasonable?

**Comment by stgraber** [comment]
Still feels a bit unneeded since the code you added to the existing patch will effectively only be useful to users who have a broken LXD right now (patch failed to apply) but those same users will also run the new patch right afterwards so kinda feels like code duplication to me.

**Comment by brauner** [comment]
It needs adaption anyway,  I think, since some of the keys have changed since (e.g. `volume.lvm.thinpool_name` etc. have been removed in newer version). So people running a broken upgrade again should use the new keys and not the old ones.

**Comment by stgraber** [comment]
Yeah, that's fine, though not technically needed since the next patch would fix those bad values for them anyway :)

File: lxc/storage.go (hunk 2 above)
Comment (brauner) [inline lxc/storage.go]: Right, I was going for the other approach of always storing bytes. :) But yeah, I will change that.

**Comment by brauner** [comment]
Closes #2931.
